File: homework2/Saytam_Mishra/q4_queue.java
package homework2.Saytam_Mishra;

class Queue {
    class Node {
        int data;
        Node next;

        public Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node head;
    private Node tail;

    public int peek() { // O(1) time
        if(head == null) {
            throw new IllegalStateException("Queue is empty");
        }
        return head.data;
    }

    public void enqueue(int x) { // O(1) time
        Node newNode = new Node(x);

        if(isEmpty()) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
    }

    public int dequeue() { // O(1) time
        if(isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        } else {
            int delete = head.data;
            head = head.next;
            if(head == null) tail = null;
            return delete;
        }
    }

    public boolean isEmpty() { // O(1) time
        return head == null;
    }
}

public class q4_queue {
    public static void main(String[] args) {
        Queue myQueue = new Queue();

        myQueue.enqueue(10);
        myQueue.enqueue(20);
        System.out.println("Front element is: " + myQueue.peek());

        int val = myQueue.dequeue();
        System.out.println("Dequeued: " + val);
        System.out.println("New front element is: " + myQueue.peek());
    }
}

// Time taken - 7m 30s

File: homework2/Aditya_Prasad/q4_Queue.cpp
// Time taken : 14 mins

#include <bits/stdc++.h>
using namespace std;

struct Node{
    int data;
    Node* next;

    Node(int val){
        data = val;
        next = nullptr;
    }
};

class Queue {
private:
    Node* front;
    Node* rear;

public:
    Queue(){
        front = nullptr;
        rear = nullptr;
    }

    int peek(){
        return front->data;

    } // returns the first item in the queue. O(1) time.

    void enqueue(int x){
        Node* newNode = new Node(x);

        if(rear == nullptr){
            front = rear = newNode;
            return;
        }

        rear->next = newNode;
        rear = newNode;
    }// adds x to the back of the queue. O(1) time.


    int dequeue(){
        Node* newNode = front;
        int ans = newNode->data;

        front = front->next;

        if(front == nullptr){
            rear = nullptr;
        
        }
        
        delete newNode;
        return ans;
    }// removes and returns the first item in the queue. O(1) time.

    bool isEmpty(){
        if(front == nullptr) return true;
        else return false;
    } // returns a boolean indicating whether the queue is empty. O(1) time.
};

int main(){
    return 0;
}